src/design-pages/passkey_authentication.rst

```diff
@@ -0,0 +1,318 @@
+
+In the Active Directory case, the
+`altSecurityIdentities <https://learn.microsoft.com/en-us/windows/win32/adschema/a-altsecurityidentities>`__
+LDAP attribute is used to store the passkey mapping data. Example:
+
+* altSecurityIdentities: passkey:9S87qLk8/RxYJ3skwwYduomAM+/HDtz41N0+w/vRL6aGKJkLMsg+2OhO0E8pK5DuO1KmdK61K8PmH7jiYuOqbg==,9YE1s/f7J47h2A/DXCVFWulqoBXFzCSxcbGEBadkpSUFjwUudhPLnPUTv2qNamakXJgRYCZQ7vpS/t5zXMLnkw==
+
+Finally, IPA ``iparequireuserverification`` LDAP attribute is retrieved from
+the IPA passkey configuration. Store this in the sysdb associated with  the IPA
+domain. This attribute stores a string (on, off, default). Example:
+
+* iparequireuserverification: on
+
+User authentication with the key
+********************************
+The idea is to create a helper process similar to the p11_child. The helper
+process receives the passkey’s user key mapping data as an argument alongside
+other parameters (check
+:ref:`Configuration options<configuration-options>` for more information).
+
+This helper process heavily relies on the libfido2 library. The first steps
+are to initialize the library and get the list of keys connected to the system.
+Next the helper process loads the configuration parameters in the assert
+structure, that comprises a collection of statements, each statement has a
+mapping for a challenge, a credential, a signature, and ancillary attributes.
+Finally, the helper process opens the key device and ask it for an
+authentication according to the information provided in the
+`assert <https://developers.yubico.com/libfido2/Manuals/fido_assert_new.html>`__
+structure.
+
+Call the passkey child in the PAM responder
++++++++++++++++++++++++++++++++++++++++++++
+In ``pam_forwarder()`` check if passkey authentication is enabled (if
+pam_passkey_auth boolean option is true and pam cmd == SSS_PAM_AUTHENTICATE).
+If so, lookup the sysdb user object of the user logging in and read the sysdb
+passkey mapping value. Parse out the key-handle and public-key strings from the
+passkey mapping value. Read passkey related configuration options
+(passkey_verification, passkey_child_timeout, debug_libfido2) and set up the
+arguments (Domain, Key-handle and Public-key) for the passkey-child.
+
+Execute the passkey_child passing the arguments listed above. The passkey PIN
+is retrieved from the PAM conversation and written to the stdin of the
+forked passkey child process(Similar to ``get_p11_child_write_buffer()``).
+Check the passkey_child return code, return PAM_SUCCESS or failure based on the
+result and call ``pam_reply()``. If the credential is discoverable, then the
+passkey_child also prints the ``userId``, and the PAM responder has to
+compare it with the one provided by the LDAP server. If they match, then it can
+return PAM_SUCCESS.
```

Comment:
Comparing of userId might be better to do in passkey_child so the whole logic in in single process.

Reply:
@sumit-bose I remember you asked to do the comparison in the PAM responder, but I can't remember why. Can you enlighten us?

Reply:
Hi,

the idea was, since the userID is created as a random number, ot handle it as a second shared secret so that e.g. in the FreeIPA Kerberos case the userID is only know to the server and the token. The userID will not be send to the client by the server but only returned fom the client to the server so that the server can check if userID is the same as the stored one. Similar for the local authentication, passkey_child should not know which value is expected here.

Originally I thought that the userId will be part of the authenticator data returned by the token and hence is protected against modification by the signature. But it looks it can only be send separately. Not sure if it would need some additional protection in this case?

bye,
Sumit

Reply:
As discussed in the meeting this part of the design is fine. So I'm resolving this conversation.